File: 39-combination-sum/39-combination-sum.cpp
class Solution {
private:
    void cs(vector<vector<int>>&v , vector<int>&a ,vector<int> &c,int tar,int i)
    {
        if(i==c.size())
        {
         if(tar==0)
          {
            v.push_back(a); 
          }
            return ;
        }
        if(c[i]<=tar)
        {
            a.push_back(c[i]);
            cs(v,a,c,tar-c[i],i);
            a.pop_back();
        }
        cs(v,a,c,tar,i+1);
        
    }
public:
    vector<vector<int>> combinationSum(vector<int>& c, int tar) {
        vector<vector<int>>v;
        vector<int>a;
        cs(v,a,c,tar,0);
        return v;
    }
};

File: 39-combination-sum/39-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>>ans = new ArrayList<>();
        List<Integer>que = new ArrayList<>();
        solve(candidates,target,0,que,ans,0);
        return ans;
    }
    public void solve(int[] arr, int target,int sum ,List<Integer>que,List<List<Integer>>ans,int idx)
    {
       if(idx == arr.length)
       {
           if(target==sum)
           {
               ans.add(new ArrayList(que));
              
           }
           return ;
       }

            if(sum<=target)
            {
                que.add(arr[idx]);
                solve(arr,target,sum+arr[idx],que,ans,idx);
                que.remove(que.size()-1);
            }
             solve(arr,target,sum,que,ans,idx+1);
    
    }
    
}